backfill/backfill.go

```diff
@@ -0,0 +1,397 @@
+package backfill
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	// Blank import to register types for CBORGEN
+	_ "github.com/bluesky-social/indigo/api/bsky"
+	"github.com/bluesky-social/indigo/repo"
+	"github.com/bluesky-social/indigo/repomgr"
+	"github.com/ipfs/go-cid"
+	typegen "github.com/whyrusleeping/cbor-gen"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+	"go.opentelemetry.io/otel"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+)
+
+// Job is an interface for a backfill job
+type Job interface {
+	Repo() string
+	State() string
+	SetState(ctx context.Context, state string) error
+
+	// FlushBufferedOps calls the given callback for each buffered operation
+	// Once done it clears the buffer and marks the job as "complete"
+	// Allowing the Job interface to abstract away the details of how buffered
+	// operations are stored and/or locked
+	FlushBufferedOps(ctx context.Context, cb func(kind, path string, rec *typegen.CBORMarshaler, cid *cid.Cid) error) error
+
+	ClearBufferedOps(ctx context.Context) error
+}
+
+// Store is an interface for a backfill store which holds Jobs
+type Store interface {
+	// BufferOp buffers an operation for a job and returns true if the operation was buffered
+	// If the operation was not buffered, it returns false and an error (ErrJobNotFound or ErrJobComplete)
+	BufferOp(ctx context.Context, repo string, kind, path string, rec *typegen.CBORMarshaler, cid *cid.Cid) (bool, error)
+	GetJob(ctx context.Context, repo string) (Job, error)
+	GetNextEnqueuedJob(ctx context.Context) (Job, error)
+}
+
+// Backfiller is a struct which handles backfilling a repo
+type Backfiller struct {
+	Name               string
+	HandleCreateRecord func(ctx context.Context, repo string, path string, rec *typegen.CBORMarshaler, cid *cid.Cid) error
+	HandleUpdateRecord func(ctx context.Context, repo string, path string, rec *typegen.CBORMarshaler, cid *cid.Cid) error
+	HandleDeleteRecord func(ctx context.Context, repo string, path string) error
+	Store              Store
+
+	// Number of backfills to process in parallel
+	ParallelBackfills int
+	// Number of records to process in parallel for each backfill
+	ParallelRecordCreates int
+	// Prefix match for records to backfill i.e. app.bsky.feed.app/
+	// If empty, all records will be backfilled
+	NSIDFilter   string
+	CheckoutPath string
+
+	Logger      *zap.SugaredLogger
```

Comment:
I don't love proliferating logging libraries. We have go-log currently, and I hoped we could transition to `log/slog`. If we want to use `zap` instead we should probably commit and use it in most places.

Reply:
The slog package definitely has a few rough edges i.e.:

```
slog.Info(fmt.Sprintf("processing backfill for %s", repoDid))
```

Instead of 
```
slog.Infof("processing backfill for %s", repoDid))
```

There might be some magic you can do to make this less ugly but I haven't seen it yet.

Refactored to use slog anyhow.

Reply:
the pattern I would use with structured logging is:

```
slog.Info("processing repo backfill", "did", repoDid)
```

for the common case of errors, I think it is acceptable (?) to do this even without checking if the error is nil or not first:

```
slog.Info("doing a thing", "err", err)
```

Reply:
Slog requires you to wrap in a `slog.Attr` but I don't think there's a `slog.Attr` for errors?

i.e.

```
slog.Info("do a thing", slog.String("my_key", "my_value"))
```

Reply:
This was one of the errors that prompted me to do what I did there:
![CleanShot 2023-08-31 at 15 31 20](https://github.com/bluesky-social/indigo/assets/1617325/9207b427-7173-432f-bbc2-507731bab86f)

Reply:
I think you can use either API: with attr or without. The with attr is a bit more performant; probably a bunch more in relative terms, but very little impact in absolute terms for infrequent logging.

I used the direct style, and confirmed it worked, here: https://github.com/bluesky-social/indigo/blob/main/atproto/identity/handle.go#L137

Reply:
I might have missed something in that case and overcorrected though, it's possible that the key, value pairs are happy without an explicit `log.attr` wrapper but if I wanted to do a format string I needed to explicitly sprintf it.

Reply:
Okay I think that's a lot cleaner now